sig-azure/0018-20180711-azure-availability-zones.md

```diff
@@ -0,0 +1,238 @@
+
+These options should be configured in Azure cloud provider configure file (e.g. `/etc/kubernetes/azure.json`):
+
+```json
+{
+    ...,
+    "loadBalancerSku": "standard",
+    "excludeMasterFromStandardLB": true
+}
+```
+
+Note that with standard SKU LoadBalancer, `primaryAvailabitySetName` and `primaryScaleSetName` is not required because all available nodes (with configurable masters via `excludeMasterFromStandardLB`) are added to LoadBalancer backend pools.
+
+## Node registration
+
+When nodes are started, kubelet automatically adds labels to them with region and zone information:
+
+- Region: `failure-domain.beta.kubernetes.io/region=centralus`
+- Zone: `failure-domain.beta.kubernetes.io/zone=centralus-1`
+
+```sh
+$ kubectl get nodes --show-labels
+NAME                STATUS    AGE   VERSION    LABELS
+kubernetes-node12   Ready     6m    v1.11      failure-domain.beta.kubernetes.io/region=centralus,failure-domain.beta.kubernetes.io/zone=centralus-1,...
+```
+
+Azure cloud providers sets fault domain for label `failure-domain.beta.kubernetes.io/zone` today. With AZ enabled, we should set the node's availability zone instead. To keep backward compatibility and distinguishing from fault domain, `<region>-<AZ>` is used here.
+
+The node's zone could get by ARM API or instance metadata. This will be added in  `GetZoneByProviderID()` and `GetZoneByNodeName()`.
+
+### Get by instance metadata
+
+This method is used in kube-controller-manager.
+
+```sh
+# Instance metadata API should be upgraded to 2017-12-01.
+$ curl -H Metadata:true "http://169.254.169.254/metadata/instance/compute/zone?api-version=2017-12-01&format=text"
+2
+```
+
+### Get by Go SDK
+
+This method is used in cloud-controller-manager.
+
+No `zones` property is included in `VirtualMachineScaleSetVM` yet in Azure Go SDK (including latest 2018-04-01 compute API).
+
+We need to ask Azure Go SDK to add `zones` for `VirtualMachineScaleSetVM`. Opened the issue https://github.com/Azure/azure-sdk-for-go/issues/2183 for tracking it.
+
+> Note: there's already `zones` property in `VirtualMachineScaleSet`, `VirtualMachine` and `Disk`.
+
+## LoadBalancer and PublicIP
+
+LoadBalancer with standard SKU will be created and all available nodes (including VirtualMachines and VirtualMachineScaleSetVms, together with optional masters configured via excludeMasterFromStandardLB) are added to LoadBalancer backend pools.
+
+PublicIPs will also be created with standard SKU, and they are zone redundant by default.
+
+Note that zonal PublicIPs are not supported. We may add this easily if there’re clear use-cases in the future.
+
+## AzureDisk
+
+When Azure managed disks are created, the `PersistentVolumeLabel` admission controller automatically adds zone labels to them. The scheduler (via `VolumeZonePredicate`) will then ensure that pods that claim a given volume are only placed into the same zone as that volume, as volumes cannot be attached across zones.
```

Comment:
Eventually the VolumeZone predicate will be deprecated and replaced by PV.NodeAffinity